File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.util.*;
import java.io.*;


public class Main {
    static final BufferedReader BR = new BufferedReader(new InputStreamReader(System.in));
    static final int[] DX = {-1, 1, 0, 0};
    static final int[] DY = {0, 0, -1, 1};
    int[] inputArray;
    int N, L, R;
    int[][] population;
    boolean[][] visited;
    int days;

    public static void main(String[] args) {

        Main main = new Main();
        try {
            main.init();
            main.solution();
        } catch (IOException e) {
            System.out.println("Exception during I/O");
        }

    }

    int[] getInputArray() throws IOException {
        return Arrays.stream(BR.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
    }

    void init() throws IOException {
        inputArray = getInputArray();
        N = inputArray[0];
        L = inputArray[1];
        R = inputArray[2];

        population = new int[N][N];

        for (int i = 0; i < N; i++) {
            population[i] = getInputArray();
        }
        days = 0;
    }


    void solution() throws IOException {
        while (true) {
            visited = new boolean[N][N];
            List<Neighbor> unions = new ArrayList<>();
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (!visited[i][j]) {
                        visited[i][j] = true;

                        Neighbor neighbors = getNeighbors(i, j);

                        if (neighbors.cities.size() > 1) {
                            unions.add(neighbors);
                        }

                    }
                }
            }
            if (unions.isEmpty()) {
                break;
            }
            updateUnion(unions);
            days += 1;
        }

        System.out.println(days);
    }

    Neighbor getNeighbors(int x, int y) {
        List<Position> cities = new ArrayList<>();
        int populationSum = 0;
        List<Position> current = Arrays.asList(new Position(x, y));
        while (!current.isEmpty()) {
            List<Position> temp = new ArrayList<>();
            for (Position position : current) {
                cities.add(position);
                populationSum += population[position.x][position.y];
                for (int i = 0; i < 4; i++) {
                    if (canGo(position.x, position.y, i)) {
                        visited[position.x + DX[i]][position.y + DY[i]] = true;
                        temp.add(new Position(position.x + DX[i], position.y + DY[i]));
                    }
                }
            }
            current = temp;
        }

        return new Neighbor(cities, populationSum);

    }

    void updateUnion(List<Neighbor> neighbors) {
        for (Neighbor neighbor : neighbors) {
            int newPopulation = neighbor.populationSum / neighbor.cities.size();
            for (Position city : neighbor.cities) {
                population[city.x][city.y] = newPopulation;
            }
        }
    }

    static class Position {
        int x;
        int y;

        public Position(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static class Neighbor {
        List<Position> cities;
        int populationSum;

        public Neighbor(List<Position> cities, int populationSum) {
            this.cities = cities;
            this.populationSum = populationSum;
        }
    }


    boolean canGo(int x, int y, int direction) {
        int newX = x + DX[direction];
        int newY = y + DY[direction];
        if (!isInner(newX, newY)) {
            return false;
        }
        int populationDiff = Math.abs(population[newX][newY] - population[x][y]);
        return L <= populationDiff && populationDiff <= R && !visited[newX][newY];
    }

    boolean isInner(int x, int y) {
        return 0 <= x && x < N && 0 <= y && y < N;
    }
}

File: 백준/Gold/16234. 인구 이동/인구 이동.py
import sys


def get_neighbor_2(i, j):
    temp = []
    if i + 1 < N and visited[i + 1][j] == 0:
        bottom = abs(population[i][j] - population[i + 1][j])
        if L <= bottom and bottom <= R:
            temp.append([i + 1, j])
    if j + 1 < N and visited[i][j + 1] == 0:
        right = abs(population[i][j] - population[i][j + 1])
        if L <= right and right <= R:
            temp.append([i, j + 1])
    return temp


def get_neighbor_4(i, j):
    temp = []
    if i - 1 >= 0 and visited[i - 1][j] == 0:
        above = abs(population[i][j] - population[i - 1][j])
        if L <= above and above <= R:
            temp.append([i - 1, j])
    if i + 1 < N and visited[i + 1][j] == 0:
        bottom = abs(population[i][j] - population[i + 1][j])
        if L <= bottom and bottom <= R:
            temp.append([i + 1, j])
    if j - 1 >= 0 and visited[i][j - 1] == 0:
        left = abs(population[i][j] - population[i][j - 1])
        if L <= left and left <= R:
            temp.append([i, j - 1])
    if j + 1 < N and visited[i][j + 1] == 0:
        right = abs(population[i][j] - population[i][j + 1])
        if L <= right and right <= R:
            temp.append([i, j + 1])
    return temp


def bfs(a, b):
    result = [[a, b]]
    go = [[a, b]]
    while go:
        temp = []
        for i, j in go:
            for k, l in get_neighbor_4(i, j):
                temp.append([k, l])
                result.append([k, l])
                visited[k][l] = 1
        go = temp
    return result


N, L, R = map(int, sys.stdin.readline().split())

population = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]
result = 0
while True:
    # 인구 이동이 필요한지 확인 -> 연합 생성
    # 연합 생성 안됐으면 loop 나감
    # 연합 생성됐으면 인구수 맞추기
    visited = [[0 for _ in range(N)] for _ in range(N)]
    unions = []
    for i in range(N):
        for j in range(N):
            if visited[i][j] == 0 and len(get_neighbor_2(i, j)) > 0:
                visited[i][j] = 1
                unions.append(bfs(i, j))

    if len(unions) == 0:
        break

    result += 1
    for cities in unions:
        # print(cities)
        city_sum = 0
        for i, j in cities:
            city_sum += population[i][j]
        city_sum = int(city_sum / len(cities))
        for i, j in cities:
            population[i][j] = city_sum
    # print(population)

print(result)
